extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java

```diff
@@ -0,0 +1,498 @@
+  }
+
+  private static List<OrderByColumnSpec> getOrderBySpec(PartitionsSpec partitionSpec)
+  {
+    if (partitionSpec.getType() == SecondaryPartitionType.RANGE) {
+      List<String> dimensions = ((DimensionRangePartitionsSpec) partitionSpec).getPartitionDimensions();
+      return dimensions.stream()
+                       .map(dim -> new OrderByColumnSpec(dim, OrderByColumnSpec.Direction.ASCENDING))
+                       .collect(Collectors.toList());
+    }
+    return Collections.emptyList();
+  }
+
+  private static Query<?> buildScanQuery(CompactionTask compactionTask, Interval interval, DataSchema dataSchema)
+  {
+    RowSignature rowSignature = getRowSignature(dataSchema);
+    return new Druids.ScanQueryBuilder().dataSource(dataSchema.getDataSource())
+                                        .columns(rowSignature.getColumnNames())
+                                        .columnTypes(rowSignature.getColumnTypes())
+                                        .intervals(new MultipleIntervalSegmentSpec(Collections.singletonList(interval)))
+                                        .legacy(false)
+                                        .filters(dataSchema.getTransformSpec().getFilter())
+                                        .context(compactionTask.getContext())
+                                        .build();
+  }
+
+  private static boolean isGroupBy(DataSchema dataSchema)
+  {
+    return dataSchema.getAggregators().length > 0;
+  }
+
+  private static boolean isQueryGranularityEmptyOrNone(DataSchema dataSchema)
+  {
+    return dataSchema.getGranularitySpec() == null
+           || dataSchema.getGranularitySpec().getQueryGranularity() == null
+           || Objects.equals(
+        dataSchema.getGranularitySpec().getQueryGranularity(),
+        Granularities.NONE
+    );
+  }
+
+  private static VirtualColumns getVirtualColumns(DataSchema dataSchema)
+  {
+    VirtualColumns virtualColumns = VirtualColumns.EMPTY;
+
+    if (!isQueryGranularityEmptyOrNone(dataSchema) && !dataSchema.getGranularitySpec()
+                                                                 .getQueryGranularity()
+                                                                 .equals(Granularities.ALL)) {
+      PeriodGranularity periodQueryGranularity = (PeriodGranularity) dataSchema.getGranularitySpec()
+                                                                               .getQueryGranularity();
+      // Need to create a virtual column for time as that's the only way to support query granularity
+      VirtualColumn virtualColumn = new ExpressionVirtualColumn(
+          TIME_VIRTUAL_COLUMN,
+          StringUtils.format(
+              "timestamp_floor(\"%s\", '%s')",
+              ColumnHolder.TIME_COLUMN_NAME,
+              periodQueryGranularity.getPeriod().toString()
+          ),
+          ColumnType.LONG,
+          new ExprMacroTable(Collections.singletonList(new TimestampFloorExprMacro()))
```

Comment:
Can't we `@JacksonInject` this? It seems weird that we are creating an expr_macro table.

Reply:
We create one even inside `QueryKitUtils#makeSegmentGranularityVirtualColumn`, so this should be fine?